In [1]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

#  Association rule mining

## Part I : Market Basket Analysis

In [2]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
import mlxtend as ml

In [3]:
mdata1 = pd.read_csv("C:/Data/Data for arules 1.csv")
mdata2 = pd.read_csv("C:/Data/Data for arules 2.csv")

In [4]:
mdata2

Row Labels  Almonds  Poha  Banana  Beans  Besan  Boiled Rice  Bread  \
0      6468572      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
1      6486475      NaN   NaN     NaN    NaN    NaN          1.0    NaN   
2      6504964      1.0   NaN     NaN    NaN    NaN          1.0    NaN   
3      6529569      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
4      6549521      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
5      6562518      NaN   NaN     NaN    1.0    NaN          1.0    NaN   
6      6596905      NaN   NaN     1.0    1.0    NaN          1.0    NaN   
7      6612806      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
8      6624000      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
9      6638454      NaN   NaN     1.0    1.0    NaN          1.0    NaN   
10     6656240      NaN   NaN     1.0    1.0    NaN          NaN    NaN   
11     6672349      NaN   NaN     NaN    1.0    NaN          1.0    NaN   
12     6687487      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
13     6704782      NaN   NaN     NaN    NaN    NaN          1.0    NaN   
14     6723554      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
15     6734605      NaN   NaN     NaN    1.0    NaN          NaN    NaN   
16     6743451      NaN   NaN     NaN    NaN    NaN          NaN    NaN   
17     6759502      NaN   NaN     NaN    NaN    NaN          NaN    NaN   
18     6774006      NaN   1.0     1.0    NaN    1.0          1.0    NaN   
19     6786934      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
20     6794402      1.0   NaN     1.0    1.0    NaN          1.0    NaN   
21     6811376      NaN   NaN     1.0    1.0    NaN          NaN    NaN   
22     7874962      NaN   NaN     1.0    1.0    NaN          NaN    NaN   
23     7887787      NaN   NaN     1.0    NaN    NaN          1.0    NaN   
24     7931131      1.0   NaN     1.0    1.0    NaN          1.0    NaN   
25     7948600      NaN   NaN     1.0    1.0    NaN          1.0    NaN   
26     7969005      NaN   NaN     1.0    NaN    NaN          NaN    1.0   
27     7980335      NaN   NaN     1.0    NaN    NaN          1.0    NaN   
28     7998535      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
29     8008269      NaN   NaN     1.0    1.0    NaN          1.0    NaN   
30     8014606      NaN   NaN     NaN    NaN    NaN          NaN    NaN   
31     8026271      NaN   NaN     1.0    1.0    1.0          NaN    NaN   
32     8030611      NaN   NaN     NaN    NaN    NaN          NaN    NaN   
33     8039388      NaN   NaN     1.0    1.0    NaN          NaN    1.0   
34     8044079      NaN   NaN     1.0    NaN    NaN          1.0    NaN   
35     8052390      NaN   NaN     NaN    1.0    NaN          NaN    NaN   
36     8057553      NaN   NaN     1.0    NaN    NaN          1.0    NaN   
37     8062224      NaN   NaN     1.0    1.0    NaN          NaN    NaN   
38     8068900      NaN   NaN     1.0    NaN    NaN          NaN    1.0   
39     8087495      NaN   NaN     1.0    1.0    NaN          NaN    NaN   
40     8093634      NaN   NaN     NaN    NaN    NaN          NaN    NaN   
41     8104219      NaN   NaN     1.0    1.0    NaN          NaN    1.0   
42     8116554      1.0   NaN     1.0    1.0    NaN          1.0    1.0   
43     8132674      1.0   NaN     1.0    1.0    NaN          1.0    1.0   
44     8143489      NaN   NaN     1.0    1.0    NaN          NaN    NaN   
45     8156232      1.0   1.0     1.0    1.0    NaN          1.0    1.0   
46     8166849      1.0   NaN     1.0    1.0    NaN          1.0    NaN   
47     8181410      NaN   NaN     NaN    1.0    NaN          NaN    NaN   
48     8184240      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
49     8199070      NaN   NaN     NaN    NaN    NaN          NaN    NaN   

    Brinjals  Buns & Pavs  ...  Sooji & Rava  Soya Products  Sugar  \
0        NaN          NaN  ...           NaN            NaN    1.0   
1        NaN          1.0  ...        

In [5]:
mdata2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 54 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Row Labels                  50 non-null     int64  
 1   Almonds                     7 non-null      float64
 2   Poha                        2 non-null      float64
 3   Banana                      36 non-null     float64
 4   Beans                       24 non-null     float64
 5   Besan                       2 non-null      float64
 6   Boiled Rice                 20 non-null     float64
 7   Bread                       7 non-null      float64
 8   Brinjals                    2 non-null      float64
 9   Buns & Pavs                 1 non-null      float64
 10  Cakes                       6 non-null      float64
 11  Cashews                     10 non-null     float64
 12  Chips                       2 non-null      float64
 13  Cookies                     2 non-nul

In [6]:
mdata2 = mdata2.fillna(0)

Checking for null values

In [7]:
mdata2.isnull().sum(axis=0)

Row Labels                    0
Almonds                       0
Poha                          0
Banana                        0
Beans                         0
Besan                         0
Boiled Rice                   0
Bread                         0
Brinjals                      0
Buns & Pavs                   0
Cakes                         0
Cashews                       0
Chips                         0
Cookies                       0
Corn Snacks                   0
Cream Biscuits                0
Eggs                          0
Ghee                          0
Gourd & Cucumber              0
Ground Coffee                 0
Hair Oil                      0
Health Drinks                 0
Healthy Snacks                0
Instant Noodles               0
Jaggery                       0
Liquid Soaps & Bars           0
Moong Dal                     0
Mosquito Repellent            0
Namkeen                       0
Organic F&V                   0
Other Dals                    0
Other Dr

In [8]:
mdata2.drop('Row Labels',axis=1,inplace=True)

Checking for unique values, our target is to get only two unique values in each column.

In [9]:
print(mdata2.nunique())

Almonds                       2
Poha                          2
Banana                        2
Beans                         2
Besan                         2
Boiled Rice                   2
Bread                         2
Brinjals                      2
Buns & Pavs                   2
Cakes                         2
Cashews                       2
Chips                         2
Cookies                       2
Corn Snacks                   2
Cream Biscuits                2
Eggs                          2
Ghee                          2
Gourd & Cucumber              2
Ground Coffee                 2
Hair Oil                      2
Health Drinks                 2
Healthy Snacks                2
Instant Noodles               2
Jaggery                       2
Liquid Soaps & Bars           2
Moong Dal                     2
Mosquito Repellent            2
Namkeen                       3
Organic F&V                   2
Other Dals                    3
Other Dry Fruits              2
Other Ri

Replacing all the values to 0 and 1

In [10]:
print(mdata2['Namkeen'].value_counts())
print(mdata2['Other Dals'].value_counts())

mdata2 = mdata2.replace(3.0, 1.0)
mdata2 = mdata2.replace(2.0, 1.0)

print(mdata2['Namkeen'].value_counts())
print(mdata2['Other Dals'].value_counts())

0.0    29
1.0    20
3.0     1
Name: Namkeen, dtype: int64
0.0    25
1.0    24
2.0     1
Name: Other Dals, dtype: int64
0.0    29
1.0    21
Name: Namkeen, dtype: int64
1.0    25
0.0    25
Name: Other Dals, dtype: int64


Converting all the values to boolean values

In [11]:
mdata2 = mdata2.replace(1.0,True)
mdata2 = mdata2.replace(0.0, False)

In [12]:
mdata2

Almonds   Poha  Banana  Beans  Besan  Boiled Rice  Bread  Brinjals  \
0     False  False    True  False  False        False  False     False   
1     False  False   False  False  False         True  False     False   
2      True  False   False  False  False         True  False     False   
3     False  False    True  False  False        False  False     False   
4     False  False    True  False  False        False  False     False   
5     False  False   False   True  False         True  False     False   
6     False  False    True   True  False         True  False     False   
7     False  False    True  False  False        False  False     False   
8     False  False    True  False  False        False  False     False   
9     False  False    True   True  False         True  False     False   
10    False  False    True   True  False        False  False     False   
11    False  False   False   True  False         True  False     False   
12    False  False    True  False  False        False  False     False   
13    False  False   False  False  False         True  False     False   
14    False  False    True  False  False        False  False     False   
15    False  False   False   True  False        False  False     False   
16    False  False   False  False  False        False  False     False   
17    False  False   False  False  False        False  False     False   
18    False   True    True  False   True         True  False     False   
19    False  False    True  False  False        False  False     False   
20     True  False    True   True  False         True  False     False   
21    False  False    True   True  False        False  False     False   
22    False  False    True   True  False        False  False     False   
23    False  False    True  False  False         True  False     False   
24     True  False    True   True  False         True  False     False   
25    False  False    True   True  False         True  False     False   
26    False  False    True  False  False        False   True     False   
27    False  False    True  False  False         True  False     False   
28    False  False    True  False  False        False  False     False   
29    False  False    True   True  False         True  False     False   
30    False  False   False  False  False        False  False     False   
31    False  False    True   True   True        False  False     False   
32    False  False   False  False  False        False  False     False   
33    False  False    True   True  False        False   True     False   
34    False  False    True  False  False         True  False     False   
35    False  False   False   True  False        False  False     False   
36    False  False    True  False  False         True  False     False   
37    False  False    True   True  False        False  False     False   
38    False  False    True  False  False        False   True     False   
39    False  False    True   True  False        False  False     False   
40    False  False   False  False  False        False  False     False   
41    False  False    True   True  False        False   True     False   
42     True  False    True   True  False         True   True      True   
43     True  False    True   True  False         True   True     False   
44    False  False    True   True  False        False  False      True   
45     True   True    True   True  False         True   True     False   
46     True  False    True   True  False         True  False     False   
47    False  False   False   True  False        False  False     False   
48    False  False    True  False  False        False  False     False   
49    False  False   False  False  False        False  False     False   

    Buns & Pavs  Cakes  ...  Sooji & Rava  Soya Products  Sugar  Sugar Cubes  \
0         False  False  ...         False          False   True         True   
1          True   True  ...         False          False  False      

### Model Training

In [13]:
apriori(mdata2, min_support=0.30)

support    itemsets
0      0.72         (2)
1      0.48         (3)
2      0.40         (5)
3      0.42        (27)
4      0.50        (29)
5      0.40        (31)
6      0.38        (34)
7      0.30        (36)
8      0.44        (39)
9      0.40        (42)
10     0.50        (45)
11     0.40        (46)
12     0.36        (49)
13     0.30        (52)
14     0.38      (2, 3)
15     0.30      (2, 5)
16     0.38     (2, 27)
17     0.36     (2, 29)
18     0.32     (2, 34)
19     0.40     (2, 39)
20     0.36     (2, 42)
21     0.38     (2, 45)
22     0.32     (2, 46)
23     0.32     (3, 29)
24     0.32     (34, 3)
25     0.34     (3, 39)
26     0.34     (3, 45)
27     0.32     (29, 5)
28     0.30    (29, 39)
29     0.36    (45, 29)
30     0.30    (49, 29)
31     0.30    (45, 39)
32     0.34    (49, 45)
33     0.30  (2, 3, 39)

The numbers written in the itemset column in the table represent the products (0-51). Item 0 refers to Almonds, item number 1 refers to Poha, item number 51 refers to Whole Spices.

In [14]:
print("Number of Rules:", len(apriori(mdata2, min_support=0.30)))

Number of Rules: 34


Now, using the use_colnames=True parameter within the apriori algorithm, we switch from items(products) numbers to item(product) names.

In [15]:
frequent_itemsets = apriori(mdata2, min_support=0.30, use_colnames=True)
frequent_itemsets

support                          itemsets
0      0.72                          (Banana)
1      0.48                           (Beans)
2      0.40                     (Boiled Rice)
3      0.42                         (Namkeen)
4      0.50                      (Other Dals)
5      0.40             (Other Rice Products)
6      0.38                (Other Vegetables)
7      0.30                         (Raisins)
8      0.44                 (Root Vegetables)
9      0.40                     (Snacky Nuts)
10     0.50                           (Sugar)
11     0.40                     (Sugar Cubes)
12     0.36                        (Urad Dal)
13     0.30                    (Whole Spices)
14     0.38                   (Banana, Beans)
15     0.30             (Banana, Boiled Rice)
16     0.38                 (Banana, Namkeen)
17     0.36              (Banana, Other Dals)
18     0.32        (Banana, Other Vegetables)
19     0.40         (Banana, Root Vegetables)
20     0.36             (Banana, Snacky Nuts)
21     0.38                   (Banana, Sugar)
22     0.32             (Banana, Sugar Cubes)
23     0.32               (Beans, Other Dals)
24     0.32         (Other Vegetables, Beans)
25     0.34          (Root Vegetables, Beans)
26     0.34                    (Beans, Sugar)
27     0.32         (Boiled Rice, Other Dals)
28     0.30     (Root Vegetables, Other Dals)
29     0.36               (Sugar, Other Dals)
30     0.30            (Other Dals, Urad Dal)
31     0.30          (Root Vegetables, Sugar)
32     0.34                 (Sugar, Urad Dal)
33     0.30  (Banana, Beans, Root Vegetables)

In the table above, it is seen that single, double and triple itemsets are formed. After we set the min_support value (0.30) and create rules sets, we create the Association Rules table according to the metric we are interested in (confidence, lift, conviction and etc.). Here we chose Confidence as the metric and its value 0.3 (30%).

In [16]:
rules1 = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.30)

In [17]:
print("Number of Rules : ", len(rules1))

Number of Rules :  44


In [18]:
rules1 = rules1.sort_values(['confidence'], ascending=False)
rules1

antecedents                consequents  antecedent support  \
37                 (Urad Dal)                    (Sugar)                0.36   
11          (Root Vegetables)                   (Banana)                0.44   
5                   (Namkeen)                   (Banana)                0.42   
13              (Snacky Nuts)                   (Banana)                0.40   
40   (Root Vegetables, Beans)                   (Banana)                0.34   
20         (Other Vegetables)                    (Beans)                0.38   
9          (Other Vegetables)                   (Banana)                0.38   
33                 (Urad Dal)               (Other Dals)                0.36   
26              (Boiled Rice)               (Other Dals)                0.40   
17              (Sugar Cubes)                   (Banana)                0.40   
1                     (Beans)                   (Banana)                0.48   
38            (Banana, Beans)          (Root Vegetables)                0.38   
22          (Root Vegetables)                    (Beans)                0.44   
15                    (Sugar)                   (Banana)                0.50   
39  (Banana, Root Vegetables)                    (Beans)                0.40   
3               (Boiled Rice)                   (Banana)                0.40   
7                (Other Dals)                   (Banana)                0.50   
31               (Other Dals)                    (Sugar)                0.50   
30                    (Sugar)               (Other Dals)                0.50   
23                    (Beans)          (Root Vegetables)                0.48   
24                    (Beans)                    (Sugar)                0.48   
34          (Root Vegetables)                    (Sugar)                0.44   
28          (Root Vegetables)               (Other Dals)                0.44   
43          (Root Vegetables)            (Banana, Beans)                0.44   
25                    (Sugar)                    (Beans)                0.50   
36                    (Sugar)                 (Urad Dal)                0.50   
21                    (Beans)         (Other Vegetables)                0.48   
18                    (Beans)               (Other Dals)                0.48   
19               (Other Dals)                    (Beans)                0.50   
27               (Other Dals)              (Boiled Rice)                0.50   
42                    (Beans)  (Banana, Root Vegetables)                0.48   
29               (Other Dals)          (Root Vegetables)                0.50   
32               (Other Dals)                 (Urad Dal)                0.50   
35                    (Sugar)          (Root Vegetables)                0.50   
10                   (Banana)          (Root Vegetables)                0.72   
14                   (Banana)                    (Sugar)                0.72   
4                    (Banana)                  (Namkeen)                0.72   
0                    (Banana)                    (Beans)                0.72   
12                   (Banana)              (Snacky Nuts)                0.72   
6                    (Banana)               (Other Dals)                0.72   
8                    (Banana)         (Other Vegetables)                0.72   
16                   (Banana)              (Sugar Cubes)                0.72   
2                    (Banana)              (Boiled Rice)                0.72   
41                   (Banana)   (Root Vegetables, Beans)                0.72   

    consequent support  support  confidence      lift  leverage  conviction  
37                0.50     0.34    0.944444  1.888889    0.1600    9.000000  
11                0.72     0.40    0.909091  1.262626    0.0832    3.080000  
5                 0.72     0.38    0.904762  1.256614    0.0776    2.940000  
13                0.72     0.36    0.900000  1.250000    0.0720    2.800000  
40                0.72     

Observation: If we examine the line with ID information of 37;

    -The probability of seeing Urad Dal and Sugar items together (support) is 34% (0.34),
    -94% (0.944444) of the people (confidence) who bought the Urad Dal item probably also bought the Sugar item,
    -The sales (lift) of the Sugar item in the shopping carts containing the Urad Dal item increased by 1.88 times,
    -How much higher (leverage) 0.16 is when Urad Dal and Sugar items are purchased together than if they are purchased independently of each other,
    -We can say that Urad Dal and Sugar items are related to each other (conviction) with a value of 9.00.
Now let's add the number of items in the antecedents and consequents parts and see the results:

In [19]:
rules1["antecedent_len"] = rules1["antecedents"].apply(lambda x: len(x))
rules1["consequents_len"] = rules1["consequents"].apply(lambda x: len(x))
rules1

antecedents                consequents  antecedent support  \
37                 (Urad Dal)                    (Sugar)                0.36   
11          (Root Vegetables)                   (Banana)                0.44   
5                   (Namkeen)                   (Banana)                0.42   
13              (Snacky Nuts)                   (Banana)                0.40   
40   (Root Vegetables, Beans)                   (Banana)                0.34   
20         (Other Vegetables)                    (Beans)                0.38   
9          (Other Vegetables)                   (Banana)                0.38   
33                 (Urad Dal)               (Other Dals)                0.36   
26              (Boiled Rice)               (Other Dals)                0.40   
17              (Sugar Cubes)                   (Banana)                0.40   
1                     (Beans)                   (Banana)                0.48   
38            (Banana, Beans)          (Root Vegetables)                0.38   
22          (Root Vegetables)                    (Beans)                0.44   
15                    (Sugar)                   (Banana)                0.50   
39  (Banana, Root Vegetables)                    (Beans)                0.40   
3               (Boiled Rice)                   (Banana)                0.40   
7                (Other Dals)                   (Banana)                0.50   
31               (Other Dals)                    (Sugar)                0.50   
30                    (Sugar)               (Other Dals)                0.50   
23                    (Beans)          (Root Vegetables)                0.48   
24                    (Beans)                    (Sugar)                0.48   
34          (Root Vegetables)                    (Sugar)                0.44   
28          (Root Vegetables)               (Other Dals)                0.44   
43          (Root Vegetables)            (Banana, Beans)                0.44   
25                    (Sugar)                    (Beans)                0.50   
36                    (Sugar)                 (Urad Dal)                0.50   
21                    (Beans)         (Other Vegetables)                0.48   
18                    (Beans)               (Other Dals)                0.48   
19               (Other Dals)                    (Beans)                0.50   
27               (Other Dals)              (Boiled Rice)                0.50   
42                    (Beans)  (Banana, Root Vegetables)                0.48   
29               (Other Dals)          (Root Vegetables)                0.50   
32               (Other Dals)                 (Urad Dal)                0.50   
35                    (Sugar)          (Root Vegetables)                0.50   
10                   (Banana)          (Root Vegetables)                0.72   
14                   (Banana)                    (Sugar)                0.72   
4                    (Banana)                  (Namkeen)                0.72   
0                    (Banana)                    (Beans)                0.72   
12                   (Banana)              (Snacky Nuts)                0.72   
6                    (Banana)               (Other Dals)                0.72   
8                    (Banana)         (Other Vegetables)                0.72   
16                   (Banana)              (Sugar Cubes)                0.72   
2                    (Banana)              (Boiled Rice)                0.72   
41                   (Banana)   (Root Vegetables, Beans)                0.72   

    consequent support  support  confidence      lift  leverage  conviction  \
37                0.50     0.34    0.944444  1.888889    0.1600    9.000000   
11                0.72     0.40    0.909091  1.262626    0.0832    3.080000   
5                 0.72     0.38    0.904762  1.256614    0.0776    2.940000   
13                0.72     0.36    0.900000  1.250000    0.0720    2.800000   
40                0.72

We can do what we did for the confidence metric above for other metrics. For the lift metric as an example:

In [20]:
rules2 = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules2 = rules2.sort_values(['lift'], ascending=False)
rules2

antecedents                consequents  antecedent support  \
37                 (Urad Dal)                    (Sugar)                0.36   
36                    (Sugar)                 (Urad Dal)                0.50   
43          (Root Vegetables)            (Banana, Beans)                0.44   
38            (Banana, Beans)          (Root Vegetables)                0.38   
21                    (Beans)         (Other Vegetables)                0.48   
20         (Other Vegetables)                    (Beans)                0.38   
33                 (Urad Dal)               (Other Dals)                0.36   
32               (Other Dals)                 (Urad Dal)                0.50   
22          (Root Vegetables)                    (Beans)                0.44   
23                    (Beans)          (Root Vegetables)                0.48   
26              (Boiled Rice)               (Other Dals)                0.40   
27               (Other Dals)              (Boiled Rice)                0.50   
42                    (Beans)  (Banana, Root Vegetables)                0.48   
39  (Banana, Root Vegetables)                    (Beans)                0.40   
31               (Other Dals)                    (Sugar)                0.50   
30                    (Sugar)               (Other Dals)                0.50   
25                    (Sugar)                    (Beans)                0.50   
24                    (Beans)                    (Sugar)                0.48   
35                    (Sugar)          (Root Vegetables)                0.50   
29               (Other Dals)          (Root Vegetables)                0.50   
28          (Root Vegetables)               (Other Dals)                0.44   
34          (Root Vegetables)                    (Sugar)                0.44   
19               (Other Dals)                    (Beans)                0.50   
18                    (Beans)               (Other Dals)                0.48   
10                   (Banana)          (Root Vegetables)                0.72   
11          (Root Vegetables)                   (Banana)                0.44   
5                   (Namkeen)                   (Banana)                0.42   
4                    (Banana)                  (Namkeen)                0.72   
13              (Snacky Nuts)                   (Banana)                0.40   
12                   (Banana)              (Snacky Nuts)                0.72   
40   (Root Vegetables, Beans)                   (Banana)                0.34   
41                   (Banana)   (Root Vegetables, Beans)                0.72   
8                    (Banana)         (Other Vegetables)                0.72   
9          (Other Vegetables)                   (Banana)                0.38   
17              (Sugar Cubes)                   (Banana)                0.40   
16                   (Banana)              (Sugar Cubes)                0.72   
1                     (Beans)                   (Banana)                0.48   
0                    (Banana)                    (Beans)                0.72   
15                    (Sugar)                   (Banana)                0.50   
14                   (Banana)                    (Sugar)                0.72   
2                    (Banana)              (Boiled Rice)                0.72   
3               (Boiled Rice)                   (Banana)                0.40   
6                    (Banana)               (Other Dals)                0.72   
7                (Other Dals)                   (Banana)                0.50   

    consequent support  support  confidence      lift  leverage  conviction  
37                0.50     0.34    0.944444  1.888889    0.1600    9.000000  
36                0.36     0.34    0.680000  1.888889    0.1600    2.000000  
43                0.38     0.30    0.681818  1.794258    0.1328    1.948571  
38                0.44     0.30    0.789474  1.794258    0.1328    2.660000  
21                0.38     

Filtering for Generated Rule Sets

Filter 1: Let's see the records with an Antecident item length of greater than 1 and a Confidence value greater than 0.50 and a Lift value greater than 1

In [21]:
rules1[(rules1['antecedent_len'] > 1) &
       (rules1['confidence'] >= 0.50) &
       (rules1['lift'] > 1) ].sort_values(['confidence'], ascending=False)

antecedents        consequents  antecedent support  \
40   (Root Vegetables, Beans)           (Banana)                0.34   
38            (Banana, Beans)  (Root Vegetables)                0.38   
39  (Banana, Root Vegetables)            (Beans)                0.40   

    consequent support  support  confidence      lift  leverage  conviction  \
40                0.72      0.3    0.882353  1.225490    0.0552        2.38   
38                0.44      0.3    0.789474  1.794258    0.1328        2.66   
39                0.48      0.3    0.750000  1.562500    0.1080        2.08   

    antecedent_len  consequents_len  
40               2                1  
38               2                1  
39               2                1

Observation: If we examine the line with ID information of 38;

    -The probability of seeing Root Vegetables, Beans and Bananas items together (support) is 30% (0.30),
    -88% (0.882353) of the people (confidence) who bought the Root Vegetables & Beans item probably also bought the Banana item,
    -The sales (lift) of the Banana item in the shopping carts containing the Root Vegetables & Beans item increased by 1.22 times,


Filter 2: Similarly, the records with the Antecedents item name Sugar, sorted by Confidence metric [Z-A]:

In [22]:
rules1[rules1['antecedents'] == {'Sugar'}].sort_values(['confidence'], ascending=False)

antecedents        consequents  antecedent support  consequent support  \
15     (Sugar)           (Banana)                 0.5                0.72   
30     (Sugar)       (Other Dals)                 0.5                0.50   
25     (Sugar)            (Beans)                 0.5                0.48   
36     (Sugar)         (Urad Dal)                 0.5                0.36   
35     (Sugar)  (Root Vegetables)                 0.5                0.44   

    support  confidence      lift  leverage  conviction  antecedent_len  \
15     0.38        0.76  1.055556      0.02    1.166667               1   
30     0.36        0.72  1.440000      0.11    1.785714               1   
25     0.34        0.68  1.416667      0.10    1.625000               1   
36     0.34        0.68  1.888889      0.16    2.000000               1   
35     0.30        0.60  1.363636      0.08    1.400000               1   

    consequents_len  
15                1  
30                1  
25                1  
36                1  
35                1

Observation: On examining all the above records; 
    
    We can say that out of 60%(0.60) of the people (confidence) who bought the Sugar item probably also bought either Banana, Other Dals, Beans, Urad Dall or Root Vegetables.

## Part II : Market Basket Analysis of Productdata

In [23]:
mdata1

resid  familysz  chb18  cha18 borvita milo zandu dred dblue prtnex holik  \
0     1.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
1     2.0       6.0    0.0    2.0      no   no   yes   no   yes     no    no   
2     3.0       4.0    1.0    1.0      no   no    no   no    no    yes   yes   
3     4.0       4.0    0.0    2.0     yes  yes    no   no    no     no   yes   
4     5.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
5     6.0       4.0    0.0    2.0      no   no    no   no    no    yes   yes   
6     7.0       8.0    0.0    3.0      no   no   yes  yes   yes     no    no   
7     8.0       6.0    0.0    1.0      no   no   yes   no   yes     no    no   
8     9.0       7.0    0.0    2.0      no   no    no  yes   yes     no    no   
9    10.0       4.0    2.0    0.0     yes  yes    no   no    no     no   yes   
10   11.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
11   12.0       4.0    2.0    0.0     yes  yes    no   no    no    yes   yes   
12   13.0       4.0    1.0    1.0      no   no    no   no    no    yes   yes   
13   14.0       4.0    0.0    2.0      no   no    no   no    no    yes   yes   
14   15.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
15   16.0       4.0    1.0    1.0      no   no    no   no    no    yes   yes   
16   17.0       4.0    0.0    2.0     yes  yes    no   no    no     no   yes   
17   18.0       8.0    0.0    3.0      no   no   yes  yes   yes     no    no   
18   19.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
19   20.0       4.0    2.0    0.0     yes  yes    no   no    no     no   yes   
20   21.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
21   22.0       4.0    2.0    0.0     yes  yes    no   no    no    yes   yes   
22   23.0       4.0    1.0    1.0      no   no    no   no    no    yes   yes   
23   24.0       4.0    0.0    2.0      no   no    no   no    no    yes   yes   
24   25.0       6.0    0.0    2.0      no   no   yes   no   yes     no    no   
25   26.0       8.0    0.0    3.0      no   no   yes  yes   yes     no    no   
26   27.0       6.0    0.0    1.0      no   no   yes   no   yes     no    no   
27   28.0       7.0    0.0    2.0      no   no    no  yes   yes     no    no   
28   29.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
29   30.0       4.0    1.0    1.0      no   no    no   no    no    yes   yes   
30   31.0       4.0    0.0    2.0     yes  yes    no   no    no     no   yes   
31   32.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
32   33.0       4.0    0.0    2.0      no   no    no   no    no    yes   yes   
33   34.0       4.0    2.0    0.0     yes  yes    no   no    no     no   yes   
34   35.0       3.0    1.0    0.0     yes  yes    no   no    no     no   yes   
35   36.0       4.0    2.0    0.0     yes  yes    no   no    no    yes   yes   
36   37.0       4.0    1.0    1.0      no   no    no   no    no    yes   yes   
37   38.0       4.0    0.0    2.0      no   no    no   no    no    yes   yes   
38   39.0       6.0    0.0    2.0      no   no   yes   no   yes     no    no   
39   40.0       8.0    0.0    3.0      no   no   yes  yes   yes     no    no   
40    NaN       NaN    NaN    NaN     NaN  NaN   NaN  NaN   NaN    NaN   NaN   
41    NaN       NaN    NaN    NaN     NaN  NaN   NaN  NaN   NaN    NaN   NaN   
42    NaN       NaN    NaN    NaN     NaN  NaN   NaN  NaN   NaN    NaN   NaN   

   badya complan  
0     no      no  
1    yes      no  
2     no     yes  
3     no      no  
4     no     yes  
5     no     yes  
6    yes      no  
7    yes      no  
8    yes      no  
9     no      no  
10    no      no  
11    no     yes  
12    no     yes  
13    no     yes  
14    no      no  
15    no     yes  
16    no      no  
17   yes      no  
18    no     yes  
19    no      no  
20    no      no  
21    no     yes  
22    no     yes  
23    no     yes  
24   yes

In [24]:
mdata1.drop({'resid','familysz','chb18','cha18'},axis=1,inplace=True)

In [25]:
mdata1 = mdata1.replace('yes',True)
mdata1 = mdata1.replace('no', False)
mdata1 = mdata1.dropna()

In [26]:
mdata1

borvita   milo  zandu   dred  dblue prtnex  holik  badya complan
0     True   True  False  False  False  False   True  False   False
1    False  False   True  False   True  False  False   True   False
2    False  False  False  False  False   True   True  False    True
3     True   True  False  False  False  False   True  False   False
4     True   True  False  False  False  False   True  False    True
5    False  False  False  False  False   True   True  False    True
6    False  False   True   True   True  False  False   True   False
7    False  False   True  False   True  False  False   True   False
8    False  False  False   True   True  False  False   True   False
9     True   True  False  False  False  False   True  False   False
10    True   True  False  False  False  False   True  False   False
11    True   True  False  False  False   True   True  False    True
12   False  False  False  False  False   True   True  False    True
13   False  False  False  False  False   True   True  False    True
14    True   True  False  False  False  False   True  False   False
15   False  False  False  False  False   True   True  False    True
16    True   True  False  False  False  False   True  False   False
17   False  False   True   True   True  False  False   True   False
18    True   True  False  False  False  False   True  False    True
19    True   True  False  False  False  False   True  False   False
20    True   True  False  False  False  False   True  False   False
21    True   True  False  False  False   True   True  False    True
22   False  False  False  False  False   True   True  False    True
23   False  False  False  False  False   True   True  False    True
24   False  False   True  False   True  False  False   True   False
25   False  False   True   True   True  False  False   True   False
26   False  False   True  False   True  False  False   True   False
27   False  False  False   True   True  False  False   True   False
28    True   True  False  False  False  False   True  False   False
29   False  False  False  False  False   True   True  False    True
30    True   True  False  False  False  False   True  False   False
31    True   True  False  False  False  False   True  False    True
32   False  False  False  False  False   True   True  False    True
33    True   True  False  False  False  False   True  False   False
34    True   True  False  False  False  False   True  False   False
35    True   True  False  False  False   True   True  False    True
36   False  False  False  False  False   True   True  False    True
37   False  False  False  False  False   True   True  False    True
38   False  False   True  False   True  False  False   True   False
39   False  False   True   True   True  False  False   True   False

In [27]:
print(mdata1.isnull().sum(axis=0))
print(mdata1.nunique())
print(mdata1.info())

borvita    0
milo       0
zandu      0
dred       0
dblue      0
prtnex     0
holik      0
badya      0
complan    0
dtype: int64
borvita    2
milo       2
zandu      2
dred       2
dblue      2
prtnex     2
holik      2
badya      2
complan    2
dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 40 entries, 0 to 39
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   borvita  40 non-null     object
 1   milo     40 non-null     object
 2   zandu    40 non-null     object
 3   dred     40 non-null     object
 4   dblue    40 non-null     object
 5   prtnex   40 non-null     object
 6   holik    40 non-null     object
 7   badya    40 non-null     object
 8   complan  40 non-null     object
dtypes: object(9)
memory usage: 3.1+ KB
None


In [28]:
frequent_itemsets = apriori(mdata1, min_support=0.20, use_colnames=True)
frequent_itemsets

support                  itemsets
0     0.45                 (borvita)
1     0.45                    (milo)
2    0.225                   (zandu)
3    0.275                   (dblue)
4     0.35                  (prtnex)
5    0.725                   (holik)
6    0.275                   (badya)
7    0.425                 (complan)
8     0.45           (milo, borvita)
9     0.45          (holik, borvita)
10    0.45             (holik, milo)
11   0.225            (dblue, zandu)
12   0.225            (badya, zandu)
13   0.275            (dblue, badya)
14    0.35           (holik, prtnex)
15    0.35         (complan, prtnex)
16   0.425          (holik, complan)
17    0.45    (milo, borvita, holik)
18   0.225     (dblue, badya, zandu)
19    0.35  (holik, complan, prtnex)

Here, Support means transaction containing a particular product by total number of transaction

For example, bornvita and milo are purchased 45% (0.45) more often than compare to zandu and bdlue 22%(0.225).

In the table above, it is seen that single, double and triple itemsets are formed. After we set the min_support value (0.25) and create rules sets, we create the Association Rules table according to the metric we are interested in (confidence, lift, conviction and etc.). Here we chose Confidence as the metric and its value 0.3 (30%).

In [29]:
rules1 = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.30)
rules1

antecedents        consequents  antecedent support  \
0              (milo)          (borvita)               0.450   
1           (borvita)             (milo)               0.450   
2             (holik)          (borvita)               0.725   
3           (borvita)            (holik)               0.450   
4             (holik)             (milo)               0.725   
5              (milo)            (holik)               0.450   
6             (dblue)            (zandu)               0.275   
7             (zandu)            (dblue)               0.225   
8             (badya)            (zandu)               0.275   
9             (zandu)            (badya)               0.225   
10            (dblue)            (badya)               0.275   
11            (badya)            (dblue)               0.275   
12            (holik)           (prtnex)               0.725   
13           (prtnex)            (holik)               0.350   
14          (complan)           (prtnex)               0.425   
15           (prtnex)          (complan)               0.350   
16            (holik)          (complan)               0.725   
17          (complan)            (holik)               0.425   
18    (borvita, milo)            (holik)               0.450   
19      (holik, milo)          (borvita)               0.450   
20   (holik, borvita)             (milo)               0.450   
21             (milo)   (holik, borvita)               0.450   
22          (borvita)      (holik, milo)               0.450   
23            (holik)    (borvita, milo)               0.725   
24     (dblue, badya)            (zandu)               0.275   
25     (dblue, zandu)            (badya)               0.225   
26     (badya, zandu)            (dblue)               0.225   
27            (dblue)     (badya, zandu)               0.275   
28            (badya)     (dblue, zandu)               0.275   
29            (zandu)     (dblue, badya)               0.225   
30   (holik, complan)           (prtnex)               0.425   
31    (holik, prtnex)          (complan)               0.350   
32  (complan, prtnex)            (holik)               0.350   
33            (holik)  (complan, prtnex)               0.725   
34          (complan)    (holik, prtnex)               0.425   
35           (prtnex)   (holik, complan)               0.350   

    consequent support  support  confidence      lift  leverage  conviction  
0                0.450    0.450    1.000000  2.222222  0.247500         inf  
1                0.450    0.450    1.000000  2.222222  0.247500         inf  
2                0.450    0.450    0.620690  1.379310  0.123750    1.450000  
3                0.725    0.450    1.000000  1.379310  0.123750         inf  
4                0.450    0.450    0.620690  1.379310  0.123750    1.450000  
5                0.725    0.450    1.000000  1.379310  0.123750         inf  
6                0.225    0.225    0.818182  3.636364  0.163125    4.262500  
7                0.275    0.225    1.000000  3.636364  0.163125         inf  
8                0.225    0.225    0.818182  3.636364  0.163125    4.262500  
9                0.275    0.225    1.000000  3.636364  0.163125         inf  
10               0.275    0.275    1.000000  3.636364  0.199375         inf  
11               0.275    0.275    1.000000  3.636364  0.199375         inf  
12               0.350    0.350    0.482759  1.379310  0.096250    1.256667  
13               0.725    0.350    1.000000  1.379310  0.096250         inf  
14               0.350    0.350    0.823529  2.352941  0.201250    3.683333  
15               0.425    0.350    1.000000  2.352941  0.201250         inf  
16               0.425    0.425    0.586207  1.379310  0.116875    1.389583  
17               0.725    0.425    1.000000  1.379310  0.116875         inf  
18               0.725    0.450    1.000000  1.379310  0.123750         inf  
19               0.450    0.450    1.000000  2.222222  0.247500         inf  
20 

Observation: If we examine the line with ID information of 0;

    -The probability of seeing Borvita and Milo items together (support) is 45% (0.450),
    -100% (1.0) of the people (confidence) who bought the Bornvit item probably also bought the Milo item,
    -The sales (lift) of the Milo item in the shopping carts containing the Borvita item increased by 2.22 times,
    -How much higher (leverage) 0.24 is when Bornvita and Milo items are purchased together than if they are purchased independently of each other,
    -We can say that Bornvita and Milo items are related to each other (conviction) with a value of infinity.

In [30]:
print("Number of rules : ",len(rules1))

Number of rules :  36


In [31]:
rules2 = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules2 = rules2.sort_values(['lift'], ascending=False)
rules2

antecedents        consequents  antecedent support  \
10            (dblue)            (badya)               0.275   
25     (dblue, zandu)            (badya)               0.225   
26     (badya, zandu)            (dblue)               0.225   
29            (zandu)     (dblue, badya)               0.225   
7             (zandu)            (dblue)               0.225   
9             (zandu)            (badya)               0.225   
11            (badya)            (dblue)               0.275   
24     (dblue, badya)            (zandu)               0.275   
6             (dblue)            (zandu)               0.275   
8             (badya)            (zandu)               0.275   
28            (badya)     (dblue, zandu)               0.275   
27            (dblue)     (badya, zandu)               0.275   
14          (complan)           (prtnex)               0.425   
15           (prtnex)          (complan)               0.350   
35           (prtnex)   (holik, complan)               0.350   
30   (holik, complan)           (prtnex)               0.425   
31    (holik, prtnex)          (complan)               0.350   
34          (complan)    (holik, prtnex)               0.425   
21             (milo)   (holik, borvita)               0.450   
22          (borvita)      (holik, milo)               0.450   
0              (milo)          (borvita)               0.450   
20   (holik, borvita)             (milo)               0.450   
19      (holik, milo)          (borvita)               0.450   
1           (borvita)             (milo)               0.450   
23            (holik)    (borvita, milo)               0.725   
17          (complan)            (holik)               0.425   
13           (prtnex)            (holik)               0.350   
12            (holik)           (prtnex)               0.725   
5              (milo)            (holik)               0.450   
4             (holik)             (milo)               0.725   
3           (borvita)            (holik)               0.450   
32  (complan, prtnex)            (holik)               0.350   
33            (holik)  (complan, prtnex)               0.725   
2             (holik)          (borvita)               0.725   
18    (borvita, milo)            (holik)               0.450   
16            (holik)          (complan)               0.725   

    consequent support  support  confidence      lift  leverage  conviction  
10               0.275    0.275    1.000000  3.636364  0.199375         inf  
25               0.275    0.225    1.000000  3.636364  0.163125         inf  
26               0.275    0.225    1.000000  3.636364  0.163125         inf  
29               0.275    0.225    1.000000  3.636364  0.163125         inf  
7                0.275    0.225    1.000000  3.636364  0.163125         inf  
9                0.275    0.225    1.000000  3.636364  0.163125         inf  
11               0.275    0.275    1.000000  3.636364  0.199375         inf  
24               0.225    0.225    0.818182  3.636364  0.163125    4.262500  
6                0.225    0.225    0.818182  3.636364  0.163125    4.262500  
8                0.225    0.225    0.818182  3.636364  0.163125    4.262500  
28               0.225    0.225    0.818182  3.636364  0.163125    4.262500  
27               0.225    0.225    0.818182  3.636364  0.163125    4.262500  
14               0.350    0.350    0.823529  2.352941  0.201250    3.683333  
15               0.425    0.350    1.000000  2.352941  0.201250         inf  
35               0.425    0.350    1.000000  2.352941  0.201250         inf  
30               0.350    0.350    0.823529  2.352941  0.201250    3.683333  
31               0.425    0.350    1.000000  2.352941  0.201250         inf  
34               0.350    0.350    0.823529  2.352941  0.201250    3.683333  
21               0.450    0.450    1.000000  2.222222  0.247500         inf  
22               0.450    0.450    1.000000  2.222222  0.247500         inf  
0  

Observation: If we examine the line with ID information of 10;

    -The probability of seeing Badya and Dblue items together (support) is 27% (0.275),
    -The sales (lift) of the Dblue item in the shopping carts containing the Badya item increased by 3.63 times,
    -How much higher (leverage) 0.199 is when Badya and Dblue items are purchased together than if they are purchased independently of each other,
    -We can say that Badya and Dblue items are related to each other (conviction) with a value of infinity.

In [32]:
rules2["antecedent_len"] = rules2["antecedents"].apply(lambda x: len(x))
rules2["consequents_len"] = rules2["consequents"].apply(lambda x: len(x))
rules2

antecedents        consequents  antecedent support  \
10            (dblue)            (badya)               0.275   
25     (dblue, zandu)            (badya)               0.225   
26     (badya, zandu)            (dblue)               0.225   
29            (zandu)     (dblue, badya)               0.225   
7             (zandu)            (dblue)               0.225   
9             (zandu)            (badya)               0.225   
11            (badya)            (dblue)               0.275   
24     (dblue, badya)            (zandu)               0.275   
6             (dblue)            (zandu)               0.275   
8             (badya)            (zandu)               0.275   
28            (badya)     (dblue, zandu)               0.275   
27            (dblue)     (badya, zandu)               0.275   
14          (complan)           (prtnex)               0.425   
15           (prtnex)          (complan)               0.350   
35           (prtnex)   (holik, complan)               0.350   
30   (holik, complan)           (prtnex)               0.425   
31    (holik, prtnex)          (complan)               0.350   
34          (complan)    (holik, prtnex)               0.425   
21             (milo)   (holik, borvita)               0.450   
22          (borvita)      (holik, milo)               0.450   
0              (milo)          (borvita)               0.450   
20   (holik, borvita)             (milo)               0.450   
19      (holik, milo)          (borvita)               0.450   
1           (borvita)             (milo)               0.450   
23            (holik)    (borvita, milo)               0.725   
17          (complan)            (holik)               0.425   
13           (prtnex)            (holik)               0.350   
12            (holik)           (prtnex)               0.725   
5              (milo)            (holik)               0.450   
4             (holik)             (milo)               0.725   
3           (borvita)            (holik)               0.450   
32  (complan, prtnex)            (holik)               0.350   
33            (holik)  (complan, prtnex)               0.725   
2             (holik)          (borvita)               0.725   
18    (borvita, milo)            (holik)               0.450   
16            (holik)          (complan)               0.725   

    consequent support  support  confidence      lift  leverage  conviction  \
10               0.275    0.275    1.000000  3.636364  0.199375         inf   
25               0.275    0.225    1.000000  3.636364  0.163125         inf   
26               0.275    0.225    1.000000  3.636364  0.163125         inf   
29               0.275    0.225    1.000000  3.636364  0.163125         inf   
7                0.275    0.225    1.000000  3.636364  0.163125         inf   
9                0.275    0.225    1.000000  3.636364  0.163125         inf   
11               0.275    0.275    1.000000  3.636364  0.199375         inf   
24               0.225    0.225    0.818182  3.636364  0.163125    4.262500   
6                0.225    0.225    0.818182  3.636364  0.163125    4.262500   
8                0.225    0.225    0.818182  3.636364  0.163125    4.262500   
28               0.225    0.225    0.818182  3.636364  0.163125    4.262500   
27               0.225    0.225    0.818182  3.636364  0.163125    4.262500   
14               0.350    0.350    0.823529  2.352941  0.201250    3.683333   
15               0.425    0.350    1.000000  2.352941  0.201250         inf   
35               0.425    0.350    1.000000  2.352941  0.201250         inf   
30               0.350    0.350    0.823529  2.352941  0.201250    3.683333   
31               0.425    0.350    1.000000  2.352941  0.201250         inf   
34               0.350    0.350    0.823529  2.352941  0.201250    3.683333   
21               0.450    0.450    1.000000  2.222222  0.247500         inf   
22               0.450    0.450    1.000000  2.222222  0.2475

In [33]:
rules2[(rules2['antecedent_len'] > 1) &
       (rules2['confidence'] >= 0.70) &
       (rules2['lift'] > 1) ].sort_values(['lift'], ascending=False)

antecedents consequents  antecedent support  consequent support  \
25     (dblue, zandu)     (badya)               0.225               0.275   
26     (badya, zandu)     (dblue)               0.225               0.275   
24     (dblue, badya)     (zandu)               0.275               0.225   
30   (holik, complan)    (prtnex)               0.425               0.350   
31    (holik, prtnex)   (complan)               0.350               0.425   
20   (holik, borvita)      (milo)               0.450               0.450   
19      (holik, milo)   (borvita)               0.450               0.450   
32  (complan, prtnex)     (holik)               0.350               0.725   
18    (borvita, milo)     (holik)               0.450               0.725   

    support  confidence      lift  leverage  conviction  antecedent_len  \
25    0.225    1.000000  3.636364  0.163125         inf               2   
26    0.225    1.000000  3.636364  0.163125         inf               2   
24    0.225    0.818182  3.636364  0.163125    4.262500               2   
30    0.350    0.823529  2.352941  0.201250    3.683333               2   
31    0.350    1.000000  2.352941  0.201250         inf               2   
20    0.450    1.000000  2.222222  0.247500         inf               2   
19    0.450    1.000000  2.222222  0.247500         inf               2   
32    0.350    1.000000  1.379310  0.096250         inf               2   
18    0.450    1.000000  1.379310  0.123750         inf               2   

    consequents_len  
25                1  
26                1  
24                1  
30                1  
31                1  
20                1  
19                1  
32                1  
18                1

From this analysis we can say that customers who are buying Badya and Zandu are more likely to buy Dblue. 
The sales (lift) of the Dblue item in the shopping carts containing the Badya and Zandu item increased by 3.63 times,
So we can keep Badya,Dblue and Zandu together.

And the second observation is that, the customer who are willing to buy Complain and Holik are more likely to buy Prtnex also. So we can group Complain, Holik and Prtnex.

***